src/core.c/Supply-coercers.pm6

```diff
@@ -783,19 +783,31 @@
 
         supply {
             my @values = nqp::create(Array) xx +@s;
+            my Int @counts = 0 xx +@s;
+            my $watermark = Inf;
             for @s.kv -> $index, $supply {
                 if &with {
                     whenever $supply -> \val {
                         @values[$index].push(val);
+                        my $new-count = @counts[$index]++;
                         emit( [[&with]] @values.map(*.shift) ) if all(@values);
-                        LAST { done }
+                        done if $new-count == $watermark;
```

Comment:
I think this should be `done if all(@counts) >= $watermark;` instead (and `$new-count` removed). Same in line 806.
With those changes test and spectest pass when combined with [remove-the-BlockAddWheneverAwaiter-completely](https://github.com/patrickbkr/rakudo/tree/supply-whenever-no-bawa).

Reply:
I have added this commit and this change to [rakudo/rakudo#5202](https://github.com/rakudo/rakudo/pull/5202).

Reply:
Good catch; fixed.

Reply:
Maybe I misunderstand, but a `done` always quits the supply, right? So this still quits the supply as soon as the watermark is hit twice.

Reply:
Ah, yes, the termination condition should always be `all(@counts) >= $watermark`. Please feel free to take this over; I don't have time until next week.

Reply:
Will do!